### PR TITLE
[FIX] website: fix request with inactive language

### PR DESCRIPTION
**Pull request by HydrionBurst**
For the first commit:
after https://github.com/odoo/odoo/pull/142814
when query with inactive/invalid context languages for website, an error will be
raised. For example, the best lang in `Accept Language` is inactive
The error is caused because of the website http customization which may fetch
translated field before the Http._pre_dispatch with the inactive language
This commit forces the context langauge to en_US in website http customization

For the second commit:
The env sync has been synced in the super().
When we override the method, if there is no customization (cls._frontend_pre_dispatch()) we don't need to do it again

Steps to reproduce:
- Install website_sale locally (runbot seems fine even with only website_sale?)
- Restart the server so the cache is emptied (see below note)
- Be sure to empty your cache and cookies, ideally just navigate to a new 127.0.0.X URL, it's easier
- Any page load will crash

Note: It's hard to be reproduced on runbot because if one is populating the cache of website._get_cached('user_id') with english (or any installed lang) before one is accessing the DB with a disabled lang, the error won't be faced. It's why people have reported the issue appearing from time to time but not every time.
Restarting the server will clear the cache and make the error happen all the time.

### CODE REVIEW

File: addons/test_website/tests/test_session.py [hunk 1]
```diff
@@ -1,9 +1,18 @@
...
 class TestWebsiteSession(odoo.tests.HttpCase):
 
     def test_01_run_test(self):
         self.start_tour('/', 'test_json_auth')
+
+    def test_02_inactive_session_lang(self):
+        session = self.authenticate(None, None)
+        self.env.ref('base.lang_fr').active = False
+
+        with self.subTest(case='fr saved but fr_FR disabled'):
```
Comment (Julien00859) [inline addons/test_website/tests/test_session.py:14]: the subtest isn't necessary

File: addons/test_website/tests/test_session.py [hunk 2]
```diff
@@ -1,9 +1,18 @@
...
+            session.context['lang'] = 'fr_FR'
+            odoo.http.root.session_store.save(session)
+            res = self.url_open('/test_website_sitemap')
```
Comment (Julien00859) [inline addons/test_website/tests/test_session.py:17]:
```suggestion
            res = self.url_open('/test_website_sitemap')  # any auth='public' route would do
```

File: odoo/addons/base/models/ir_http.py [hunk 3]
```diff
@@ -204,6 +199,8 @@ def _pre_dispatch(cls, rule, args):
             if not isinstance(val, models.BaseModel):
                 continue
 
+            # Replace uid and lang placeholder by the current request.env.uid and request.env.lang
+            args[key] = val.with_env(request.env)
```
Comment (Julien00859) [inline odoo/addons/base/models/ir_http.py:203]: Nitpicking, add an empty line so we understand that the comment is only about L203 :pray: 
```suggestion
            # Replace uid and lang placeholder by the current request.env.uid and request.env.lang
            args[key] = val.with_env(request.env)

```

File: odoo/addons/test_http/tests/test_session.py [hunk 4]
```diff
@@ -134,6 +134,13 @@ def test_session6_saved_lang(self):
             res = self.url_open('/test_http/echo-http-context-lang')
             self.assertEqual(res.text, 'en_US')
 
+        milky_way = self.env.ref('test_http.milky_way')
+        with self.subTest(case='fr saved but fr_FR disabled'):
```
Comment (Julien00859) [inline odoo/addons/test_http/tests/test_session.py:138]:
```suggestion
        with self.subTest(case='fr record in url but fr_FR disabled'):
```

**Review by Julien00859 — APPROVED** [review]
Looks good to me! Just a few nitpickings...

**Comment by Julien00859** [comment]
@robodoo rebase-ff

File: addons/test_website/tests/test_session.py [hunk 5]
```diff
@@ -1,9 +1,16 @@
...
 
 
 @odoo.tests.common.tagged('post_install', '-at_install')
 class TestWebsiteSession(odoo.tests.HttpCase):
 
     def test_01_run_test(self):
         self.start_tour('/', 'test_json_auth')
+
+    def test_02_inactive_session_lang(self):
+        session = self.authenticate(None, None)
```
Comment (rdeodoo) [inline addons/test_website/tests/test_session.py:11]: Shouldn't there be a test in base modules? Doesn't seems related to website? Any module could do this kind of code before predispatch no?

File: addons/http_routing/models/ir_http.py [hunk 6]
```diff
@@ -547,10 +547,10 @@ def _pre_dispatch(cls, rule, args):
...
             cls._frontend_pre_dispatch()
 
-        # update the context of "<model(...):...>" args
-        for key, val in list(args.items()):
-            if isinstance(val, models.BaseModel):
-                args[key] = val.with_context(request.context)
+            # update the context of "<model(...):...>" args
+            for key, val in list(args.items()):
+                if isinstance(val, models.BaseModel):
+                    args[key] = val.with_context(request.context)
```
Comment (rdeodoo) [inline addons/http_routing/models/ir_http.py:553]: ?? The commit message is really lacking some explanation.. 

> when not request.is_frontend
the env of args don't need to be synchronized with the request again

The most important thing is.. why? Always write one in your commit message
Right now, you will be the only one to figure why was this change needed.

**Review by rdeodoo — COMMENTED** [review]
Thanks for the fix

But there is not a single steps to reproduce in any of the commits, I just tried with only french set in my browser (and only french correctly set in accepted language header) and no cookies previously set, and didn't have the error.
Colleagues in the team reported the issue also, but didn't seem to be happening all the time, is it failing only under some circumstances?
Could you please elaborate a bit more?

image

A fix should really never be written without any step to reproduce in the commit message ideally.

File: addons/http_routing/models/ir_http.py (hunk 6 above)
Comment (HydrionBurst) [reply addons/http_routing/models/ir_http.py:553]: Becaue it has been synced in the `super()`.
When you override the method, if there is no customization (`cls._frontend_pre_dispatch()`) you don't need to do it **again**

File: addons/website/models/ir_http.py [hunk 7]
```diff
@@ -127,7 +127,7 @@ def _get_converters(cls):
     @classmethod
     def _get_public_users(cls):
         public_users = super()._get_public_users()
-        website = request.env(user=SUPERUSER_ID)['website'].get_current_website()  # sudo
+        website = request.env(user=SUPERUSER_ID)['website'].with_context(lang='en_US').get_current_website()  # sudo
```
[the remark this answers is not in the document]
Comment (HydrionBurst) [reply addons/website/models/ir_http.py:130]: In the orm `lang='en_US'` is always allowed no matter if the 'en_US' is active or not.

@Julien00859  told me, in the design of the http stack, we shouldn't do any ORM query before `_pre_dispatch` which fallbacks request.env.lang to a valid one.
Here is a hack for `webiste` so we fix it here.

File: addons/test_website/tests/test_session.py (hunk 5 above)
Comment (HydrionBurst) [reply addons/test_website/tests/test_session.py:11]: This test is only for the override of `def _get_public_users(cls):` in the website which tries to query record before `_pre_dispatch`.

**Comment by HydrionBurst** [comment]
@rdeodoo 
It can be easily reprodcue in the buggy runbot.
If you want to reproduce it locally

For the problem solved in 1st commit:
you need to install `website_sale`, because it add a new translated field `prevent_zero_price_sale_text` which will be prefetched in the `_get_public_users` of website

For the prblem solved in the 3rd commit:
you can install `website_event` and access localhost:8069/event/openwood-collection-online-reveal-8/

We already create tests to reproduce the error for each commit.

**Comment by rdeodoo** [comment]
> It can be easily reprodcued in the buggy runbot.

How? I tried on runbot "all" in master with only french and no previous cookies (so no forced lang already set eg), as you can see in the screenshot I provided, but couldn't see the error.
Is this something not appearing systematically? Did I tried to reproduce it wrongly?

> We have already created tests to reproduce the error for each commit.

No, not for second one, but that part is fine, it's just about having a written step to reproduce each time you write a fix. We can't just read all tests to figure why is the fix needed and how to reproduce it :)

File: addons/http_routing/models/ir_http.py (hunk 6 above)
Comment (rdeodoo) [reply addons/http_routing/models/ir_http.py:553]: Nice, that would really need to be part of the commit message :+1: 
No need to change here since it was r+ but remind it for future commits

Comment (HydrionBurst) [reply addons/http_routing/models/ir_http.py:553]: I will add these info the description of the PR

File: addons/website/models/ir_http.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/models/ir_http.py:130]: Aren't those kind of code also going to fail then?

https://github.com/odoo/odoo/blame/2236a10de5451cb488512c322fc004368357abaa/addons/calendar/models/ir_http.py#L14
https://github.com/odoo/odoo/blame/2236a10de5451cb488512c322fc004368357abaa/addons/mail_plugin/models/ir_http.py#L13

Seems like it's also code executed before the `_pre_dispatch` doing some ORM calls?

**Comment by rdeodoo** [comment]
For tracking purpose, after discussion on discord with @HydrionBurst :
- It's hard to be reproduced on runbot because if one is populating the cache of `website._get_cached('user_id')` with english (or any installed lang) before one is accessing the DB with a disabled lang, the error won't be faced. It's why people have reported the issue appearing from time to time but not every time.
- About the step to reproduce, locally, just install website_sale and restart your browser + clear cache/cookies.

I'll adapt the PR message

File: addons/website/models/ir_http.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (HydrionBurst) [reply addons/website/models/ir_http.py:130]: Thanks for the corner cases.
The first one seems risky I will try to trigger the error
the method used in mail_plugin pypasses ORM, so it should be safe.
